cmd/emit.go

```diff
@@ -0,0 +1,142 @@
+	emitCmd.Flags().StringP("file", "f", "", "Path to a local file containing CloudEvent data to be sent  (Env: $FUNC_FILE)")
+	emitCmd.Flags().StringP("content-type", "c", "application/json", "The MIME Content-Type for the CloudEvent data  (Env: $FUNC_CONTENT_TYPE)")
+	emitCmd.Flags().StringP("path", "p", cwd(), "Path to the project directory. Ignored when --sink is provided (Env: $FUNC_PATH)")
+}
+
+var emitCmd = &cobra.Command{
+	Use:   "emit",
+	Short: "Emit a CloudEvent to a function endpoint",
+	Long: `Emit event
+
+Emits a CloudEvent, sending it to the deployed function.
+`,
+	Example: `
+# Send a CloudEvent to the deployed function with no data and default values
+# for source, type and ID
+kn func emit
+
+# Send a CloudEvent to the deployed function with the data found in ./test.json
+kn func emit --file ./test.json
+
+# Send a CloudEvent to the function running locally with a CloudEvent containing
+# "Hello World!" as the data field, with a content type of "text/plain"
+kn func emit --data "Hello World!" --content-type "text/plain" -s local
+
+# Send a CloudEvent to the function running locally with an event type of "my.event"
+kn func emit --type my.event --sink local
+
+# Send a CloudEvent to the deployed function found at /path/to/fn with an id of "fn.test"
+kn func emit --path /path/to/fn -i fn.test
+
+# Send a CloudEvent to an arbitrary endpoint
+kn func emit --sink "http://my.event.broker.com"
+`,
+	SuggestFor: []string{"meit", "emti", "send"},
+	PreRunE:    bindEnv("source", "type", "id", "data", "file", "path", "sink", "content-type"),
+	RunE:       runEmit,
+}
+
+func runEmit(cmd *cobra.Command, args []string) (err error) {
+	config := newEmitConfig()
+	var endpoint string
+	if config.Sink != "" {
+		if config.Sink == "local" {
+			endpoint = "http://localhost:8080"
+		} else {
+			endpoint = config.Sink
+		}
+	} else {
+		var f fn.Function
+		f, err = fn.NewFunction(config.Path)
+		if err != nil {
+			return
+		}
+		// What happens if the function hasn't been deployed but they don't run with --local=true
+		// Maybe we should be thinking about saving the endpoint URL in func.yaml after each deploy
+		var d *knative.Describer
+		d, err = knative.NewDescriber("")
+		if err != nil {
+			return
+		}
+		var desc fn.Description
+		desc, err = d.Describe(f.Name)
+		if err != nil {
+			return
+		}
+		// Use the first available route
+		endpoint = desc.Routes[0]
+	}
+
+	emitter := cloudevents.NewEmitter()
+	emitter.Source = config.Source
+	emitter.Type = config.Type
+	emitter.Id = config.Id
+	emitter.ContentType = config.ContentType
+	emitter.Data = config.Data
+	if config.File != "" {
+		var buf []byte
+		if emitter.Data != "" && config.Verbose {
+			// TODO: This made me wonder whether we should switch to a real logging library
+			fmt.Printf("WARN: Found both --data and --file. Using file: %v\n", config.File)
```

Comment:
@lance I would take the purist approach and say that there is no way this is a state the developer intended, but is due to either a misconfiguration or misunderstanding, and should thus be an error.